scvi/dataset/synthetic.py

```diff
@@ -42,3 +42,113 @@ def __init__(self, mu=4.0, theta=2.0, dropout=0.7, save_path='data/'):
+            The ratio_genes_zi parameter allow to control the proportion of genes that are going
+            to be ZI
+        :param ratio_genes_zi: Proportion of ZI genes
+        """
+
+        assert mode in ['mixed', 'zi', 'nb'], 'Mode {} not recognized'.format(mode)
+        np.random.seed(0)
+
+        if n_genes_total % n_clusters > 0:
+            print("Warning, clusters have inequal sizes")
+
+        if n_genes_high > (n_genes_total // n_clusters):
+            print("Overlap of", n_genes_high - (n_genes_total // n_clusters), "genes")
+
+        # Generate data before dropout
+        batch_size = n_cells_cluster * n_clusters
+        data = np.ones((n_batches, batch_size, n_genes_total))
+
+        # For each cell cluster, some genes have a high expression, the rest
+        # has a low expression. The scope of high expression genes "moves"
+        # with the cluster
+        for cluster in range(n_clusters):
+            ind_first_gene_cluster = cluster * (n_genes_total // n_clusters)
+            ind_last_high_gene_cluster = ind_first_gene_cluster + n_genes_high
+
+            # Weights in a cluster to create highly-expressed and low-expressed genes
+            weights = weight_low * np.ones((n_genes_total,))
+            weights[ind_first_gene_cluster:ind_last_high_gene_cluster] = weight_high
+            weights /= weights.sum()
+
+            exprs = np.random.poisson(lam_0 * weights, size=(n_batches, n_cells_cluster,
+                                                             n_genes_total))
+            data[:, cluster * n_cells_cluster:(cluster + 1) * n_cells_cluster, :] = exprs
+        # Apply dropout depending on the mode
+        if mode == "zi":
+            mask = np.random.binomial(n=1, p=1 - p_dropout,
+                                      size=(n_batches, batch_size, n_genes_total))
+            newdata = (data * mask)  # We put the batch index first
+        elif mode == 'nb':
+            newdata = data
+        else:
+            assert ratio_genes_zi is not None
+            assert ratio_genes_zi <= 1.
+            n_genes_zi = int(n_genes_total * ratio_genes_zi)
+
+            submask = np.random.binomial(n=1, p=1 - p_dropout,
+                                         size=(n_batches, batch_size, n_genes_zi))
+            mask = np.ones_like(data)
+            random_permutation = np.random.permutation(np.arange(n_genes_total))
+            self.zi_genes_idx = random_permutation[:n_genes_zi]
+            mask[:, :, random_permutation[:n_genes_zi]] = submask
+
+            newdata = data * mask
+        labels = np.random.randint(0, n_labels, size=(n_batches, batch_size, 1))
+        super().__init__(
+            *GeneExpressionDataset.get_attributes_from_list(newdata, list_labels=labels),
+            gene_names=np.arange(n_genes_total).astype(np.str))
+
+
+class CorrNBDataset(SyntheticDatasetCorrelated):
+    def __init__(self):
+        super().__init__(mode='nb')
```

Comment:
If these classes just call the constructor for `SyntheticDatasetCorrelated`, I'm not sure it's worth having them. It's probably clearer just to call `SyntheticDatasetCorrelated` with the right mode directly.

Alternatively, you might remove the `mode` parameter from `SyntheticDatasetCorrelated`, keep these subclasses, and implement the mode-specific behavior in the subclasses by overriding a method in the parent class. (Rather than branching on mode.) If you're overriding methods that's a good use of subclasses. Either way seems fine and preferable to the current flow.

Reply:
Ok thanks, indeed this would be more readable. We are implementing the second option.